src/extra/components/xgm.py

```diff
@@ -0,0 +1,353 @@
+
+    def _check_sase_arg(self, sase):
+        """Helper function to return the correct PropertyGroup."""
+        if len(self._extra_sases) == 0 and sase is not None:
+            raise RuntimeError("This XGM doesn't have data from any other SASEs")
+        if sase is not None and sase not in [0, 1, 3]:
+            raise RuntimeError(f"Invalid SASE '{sase}', it must be either 0, 1, or 3")
+
+        return PropertyGroup(sase) if sase is not None else self._default_pg
+
+    def intensity(self, sase=None):
+        """Return a 2D array of the pulse intensities.
+
+        This has dimensions of `(trainId, pulseId)`. For runs with a varying
+        number of pulses, the data will be sliced to the *maximum number* of
+        pulses. e.g. if a run has 100 trains with only one train containing 10
+        pulses and all the others 0, the returned array will have a shape of
+        `(100, 10)`.
+
+        Args:
+            sase (int): Specify a SASE to retrieve data for. For XGMs from SASE
+                1 and SASE 3 this can be either `1` or `3` or `0`. Passing `0`
+                is a special case that refers to the main properties of an XGM,
+                i.e. it will look at `data.intensityTD` instead of
+                `data.intensitySa1TD` with `sase=1`. This setting overrides the
+                `default_sase` argument to the
+                [constructor][extra.components.XGM].
+        """
+        pg = self._check_sase_arg(sase)
+        if pg not in self._intensity:
+            if pg == PropertyGroup.MAIN:
+                key = "data.intensityTD"
+            else:
+                key = f"data.intensitySa{pg.value}TD"
+
+            max_pulses = self.max_pulses(sase)
+            # We could use .xarray(roi=...) here, but that fails because of h5py
+            # when max_pulses is 0.
+            intensity = self.instrument_source[key].xarray()[:, :max_pulses]
+            # Assign a pulseId dimension and coordinate
+            intensity = intensity.rename(dim_0="pulseId").assign_coords(pulseId=np.arange(max_pulses))
+            # Replace the default fill value of 1 with 0
+            intensity.data[intensity.data == 1] = 0
+
+            self._intensity[pg] = intensity
+
+        return self._intensity[pg]
+
+    def npulses(self, sase=None):
+        """Return a 1D array of the number of pulses in each train.
+
+        Args:
+            sase (int): Same meaning as in
+                [XGM.intensity()][extra.components.XGM.intensity].
+        """
+        pg = self._check_sase_arg(sase)
+        if pg not in self._npulses:
+            if pg == PropertyGroup.MAIN:
+                key = "pulseEnergy.numberOfBunchesActual"
+            else:
+                key = f"pulseEnergy.numberOfSa{pg.value}BunchesActual"
+            self._npulses[pg] = self._control_source[key].xarray()
+
+        return self._npulses[pg]
+
+    def mean_pulses(self, sase=None) -> float:
+        """The mean number of pulses in a run.
```

Comment:
```suggestion
        """The mean number of pulses.
```

Technically `DataCollection` is more correct than `run`, although maybe `run` is clearer for people? 🤔 
Maybe removing it entirely is best since it's implicit it applies to the data we've instantiated the class with?

opinions?

note: this would apply to a few other places.

Reply:
I've removed `run` from these docstrings, but I kept it for the constructor because I think it's clearer. I don't feel very strongly about it though.